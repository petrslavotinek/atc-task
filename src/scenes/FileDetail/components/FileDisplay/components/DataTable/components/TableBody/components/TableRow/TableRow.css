.TableRow td {
  padding: 0 10px;
  height: 35px;
  border: 1px solid #d5d5d5;
  vertical-align: middle; }
  .TableRow td:first-child {
    padding: 0;
    width: 35px;
    text-align: center; }

.TableRow.TableRow--odd td {
  background-color: #EEE; }

.TableRow__nested > td {
  padding: 20px; }
