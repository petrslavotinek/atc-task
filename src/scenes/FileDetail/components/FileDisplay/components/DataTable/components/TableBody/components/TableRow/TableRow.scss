@import "scenes/styles/_variables";

.TableRow {
  td {
    padding: 0 $spacing;
    height: 35px;
    border: 1px solid $darker-grey;
    vertical-align: middle;

    &:first-child {
      padding: 0;
      width: 35px;
      text-align: center;
    }
  }

  &.TableRow--odd {
    td {
      background-color: $grey;
    }
  }
}

.TableRow__nested {
  & > td {
    padding: $double-spacing;
  }
}