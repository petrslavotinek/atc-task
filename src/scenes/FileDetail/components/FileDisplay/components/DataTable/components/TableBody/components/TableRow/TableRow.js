import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { itemShape } from 'scenes/FileDetail/propTypeShapes';
import ToggleButton from './components/ToggleButton/ToggleButton';
import NestedItems from './components/NestedItems/NestedItems';
import './TableRow.css';

class TableRow extends Component {
  static propTypes = {
    headers: PropTypes.arrayOf(PropTypes.string).isRequired,
    item: itemShape.isRequired,
    even: PropTypes.bool.isRequired
  }

  state = {
    open: false
  }

  handleToggle = () => {
    const { open } = this.state;
    this.setState({ open: !open });
  }

  render() {
    const { headers, item: { data, hasKids, kids }, even } = this.props;
    const { open } = this.state;
    return (
      <React.Fragment>
        <tr className={classnames('TableRow', { 'TableRow--even': even, 'TableRow--odd': !even })}>
          <td>{hasKids && <ToggleButton open={open} onClick={this.handleToggle} />}</td>
          {headers.map(header => (
            <td key={header}>{data[header]}</td>
          ))}
        </tr>
        {open && <tr className="TableRow__nested">
          <td colSpan={headers.length + 1}>
            <NestedItems kids={kids} />
          </td>
        </tr>}
      </React.Fragment>
    );
  }
}

export default TableRow;