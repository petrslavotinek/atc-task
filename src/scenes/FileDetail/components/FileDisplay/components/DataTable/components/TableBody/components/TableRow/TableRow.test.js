import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import TableRow from './TableRow';

describe('TableRow component', () => {
  const data = {
    headers: ['prop1'],
    hasItems: true,
    items: [{
      data: {
        'prop1': 'value1'
      },
      hasKids: true,
      kids: [{
        name: 'kid1',
        headers: ['prop2'],
        items: [{
          data: {
            'prop2': 'value2'
          },
          hasKids: false,
          kids: []
        }]
      }]
    }, {
      data: {
        'prop1': 'value1'
      },
      hasKids: false,
      kids: []
    }]
  };

  beforeAll(() => {
    Object.defineProperty(React, 'Fragment', {
      configurable: true,
      value: ({ children }) => (
        <table>
          <tbody>
            {Array.isArray(children)
              ? children.map(el => el)
              : children}
          </tbody>
        </table>
      )
    });
  });

  afterAll(() => {
    Reflect.deleteProperty(React, 'Fragment');
  });

  it('should render', () => {
    expect(renderer.create(
      <TableRow
        headers={data.headers}
        item={data.items[0]}
        even={false}
      />
    ).toJSON()).toMatchSnapshot();
  });

  it('should not render toggle button when having no kids', () => {
    expect(mount(
      <TableRow
        headers={data.headers}
        item={data.items[1]}
        even={false}
      />
    ).find('td button').length).toBe(0);
  });

  it('should render toggle button when having kids', () => {
    expect(mount(
      <TableRow
        headers={data.headers}
        item={data.items[0]}
        even={false}
      />
    ).find('td button').length).toBe(1);
  });

  it('should be collapsed', () => {
    expect(mount(
      <TableRow
        headers={data.headers}
        item={data.items[0]}
        even={false}
      />
    ).find('tr').length).toBe(1);
  });

  it('should toggle', () => {
    const row = shallow(
      <TableRow
        headers={data.headers}
        item={data.items[0]}
        even={false}
      />
    );

    expect(row.find('tr').length).toBe(1);

    // open
    row.find('ToggleButton').simulate('click');
    expect(row.find('tr').length).toBe(2);

    // collapse
    row.find('ToggleButton').simulate('click');
    expect(row.find('tr').length).toBe(1);
  });
});