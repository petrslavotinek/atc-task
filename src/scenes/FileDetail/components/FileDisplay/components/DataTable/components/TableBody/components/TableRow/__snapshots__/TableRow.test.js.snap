// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow component should render 1`] = `
<table>
  <tbody>
    <tr
      className="TableRow TableRow--odd"
    >
      <td>
        <button
          className="IconButton"
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-caret-right fa-w-6 "
            data-icon="caret-right"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 192 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </td>
      <td>
        value1
      </td>
    </tr>
  </tbody>
</table>
`;
